sentry-apollo-3/src/main/java/io/sentry/apollo3/SentryApollo3HttpInterceptor.kt

```diff
@@ -0,0 +1,162 @@
+        val buffer = Buffer()
+        request.body?.writeTo(buffer)
+
+        val reader = InputStreamReader(buffer.inputStream())
+
+        val content = try {
+            hub.options.serializer.deserialize(
+                reader,
+                ApolloRequestBodyContent::class.java,
+                ApolloRequestBodyContent.Deserializer
+            )
+        } catch (e: Exception) {
+            hub.options.logger.log(SentryLevel.ERROR, "Error deserializing Apollo Request Body.", e)
+        }
+
+        val span = when (content) {
+            is ApolloRequestBodyContent -> createSpanFromBodyContent(activeSpan, request, content)
+            else -> activeSpan.startChild("apollo.request.unknown")
+        }
+
+        val operationId = request.headers.firstOrNull { it.name == "X-APOLLO-OPERATION-ID" }?.value ?: "unknown"
+
+        return span.apply {
+            setData("operationId", operationId)
+        }
+    }
+
+    private fun createSpanFromBodyContent(activeSpan: ISpan, request: HttpRequest, content: ApolloRequestBodyContent): ISpan {
+        val operationType = parseOperationType(content)
+        val operationName = content.operationName ?: request.headers.firstOrNull { it.name == "X-APOLLO-OPERATION-NAME" }?.value ?: "unknown"
+        val description = "$operationType $operationName"
+
+        return activeSpan.startChild(operationName, description).apply {
+            setData("variables", content.variables.toString())
+        }
+    }
+
+    private fun finish(span: ISpan, request: HttpRequest, response: HttpResponse? = null) {
+        var newSpan: ISpan = span
+        if (beforeSpan != null) {
+            try {
+                newSpan = beforeSpan.execute(span, request, response)
+            } catch (e: Exception) {
+                hub.options.logger.log(SentryLevel.ERROR, "An error occurred while executing beforeSpan on ApolloInterceptor", e)
+            }
+        }
+        newSpan.finish()
+
+        response?.let { httpResponse ->
+            val breadcrumb =
+                Breadcrumb.http(request.url, request.method.name, httpResponse.statusCode)
+
+            request.body?.contentLength.ifHasValidLength { contentLength ->
+                breadcrumb.setData("request_body_size", contentLength)
+            }
+
+            httpResponse.body?.peek()?.readByteArray()?.size?.toLong().ifHasValidLength { contentLength ->
+                breadcrumb.setData("response_body_size", contentLength)
+            }
+
+            val hint = Hint().also {
+                it.set(TypeCheckHint.APOLLO_REQUEST, request)
+                it.set(TypeCheckHint.APOLLO_RESPONSE, httpResponse)
+            }
+
+            hub.addBreadcrumb(breadcrumb, hint)
+        }
+    }
+
+    private fun parseOperationType(content: ApolloRequestBodyContent?): String {
+        val operationPart = content?.query?.takeWhile { !it.isWhitespace() }
+        return KNOWN_OPERATIONS.firstOrNull { it == operationPart } ?: "unknown"
```

Comment:
`SentryApolloInterceptor` fall back to the class simple name, ideally we'd follow the same pattern, `unknown` isn't useful.

Reply:
We don't have access to the High-Level Operation. So we can't use the class simple name here.
In Apollo3 they split up the interceptors into Application and HttpInterceptors.

I started with the ApplicationInterceptor but ran into a few issues that could be resolved by using the HttpInterceptor instead. For example we didn't have access to request/response body to add their sizes to the breadcrumb. Also there was no way to get the request url.

Reply:
Gotcha, I'd use the interceptor that makes more sense, give us richer values, even if that means dropping request/response size, the `url` is important only if it has something useful.

Reply:
With the `url` we can check against `TraceOrigins` and decide whether to set the trace/baggage headers.